test/metabase/lib/card_test.cljc

```diff
@@ -91,3 +94,62 @@
+      (is (=? [{:name                         "CATEGORY"
+                :lib/source                   :source/breakouts
+                :lib/source-column-alias      "CATEGORY"
+                :metabase.lib.join/join-alias "Products"
+                :lib/desired-column-alias     "Products__CATEGORY"}
+               {:name                     "count"
+                :lib/source               :source/aggregations
+                :lib/source-column-alias  "count"
+                :lib/desired-column-alias "count"}]
+              cols))
+      (is (=? [[:field {:join-alias "Products"} (meta/id :products :category)]
+               [:aggregation {} string?]]
+              (map lib.ref/ref cols))))))
+
+(deftest ^:parallel returned-columns-31769-source-card-test
+  (testing "Queries with `:source-card`s with joins should return correct column metadata/refs (#31769)"
+    (let [metadata-provider (lib.tu.mocks-31769/mock-metadata-provider)
+          card            (lib.metadata/card metadata-provider 1)
+          q                 (lib/query metadata-provider card)
+          cols              (lib/returned-columns q)]
+      (is (=? [{:name                     "CATEGORY"
+                :lib/source               :source/card
+                :lib/source-column-alias  "CATEGORY"
+                :lib/desired-column-alias "Products__CATEGORY"}
+               {:name                     "count"
+                :lib/source               :source/card
+                :lib/source-column-alias  "count"
+                :lib/desired-column-alias "count"}]
+              cols))
+      (is (=? [[:field {:base-type :type/Text} "Products__CATEGORY"]
+               [:field {:base-type :type/Integer} "count"]]
+              (map lib.ref/ref cols))))))
+
+(deftest ^:parallel returned-columns-31769-source-card-previous-stage-test
+  (testing "Queries with `:source-card`s with joins in the previous stage should return correct column metadata/refs (#31769)"
+    (let [metadata-provider (lib.tu.mocks-31769/mock-metadata-provider)
+          card            (lib.metadata/card metadata-provider 1)
+          q                 (-> (lib/query metadata-provider card)
+                                lib/append-stage)
+          cols              (lib/returned-columns q)]
+      (is (=? [{:name                     "CATEGORY"
+                :lib/source               :source/previous-stage
+                :lib/source-column-alias  "Products__CATEGORY"
+                :lib/desired-column-alias "Products__CATEGORY"}
+               {:name                     "count"
+                :lib/source               :source/previous-stage
+                :lib/source-column-alias  "count"
+                :lib/desired-column-alias "count"}]
+              cols))
+      (is (=? [[:field {:base-type :type/Text} "Products__CATEGORY"]
+               [:field {:base-type :type/Integer} "count"]]
+              (map lib.ref/ref cols))))))
```

Comment:
What's so special in adding a stage that it changes `:lib/source-column-alias`?

Reply:
It's the difference between it being `col` in the first stage

```sql
SELECT join.col AS join__col
```

vs `join__col` in the next stage

```sql
SELECT source.join_col
FROM (
  SELECT join.col AS join__col
)
```

Reply:
I think this is the change I expected to see in `:name`.